src/net_processing.cpp

```diff
@@ -906,7 +907,11 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
 }
 
 void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
+    int misbehavior{0};
     {
+        PeerRef peer = GetPeerRef(nodeid);
+        assert(peer != nullptr);
```

Comment:
This is mostly a question for future stuff that gets moved in but... What if the refcount is more than 1 here? Do we continue tearing it down?

As a contrived (bad) example: what if someone bumps the misbehavior score after it's been evaluated here?

Reply:
We'll remove the `Peer` shared ptr from `g_peer_map`, but we don't destruct the object until the refcount drops to 0. Your contrived example is fine. There would have to be another thread currently in `Misbehaving()` holding a shared ptr to the `Peer` object. We'd remove it from the map, the other thread would increment the misbehaving score, and then the `Peer` object would be destructed when that thread left the `Misbehaving()` function.

I've taken a look at the future commits in https://github.com/jnewbery/bitcoin/tree/2020-06-cs-main-split and I can't see any problems with this style of teardown.

Reply:
My point here was that calls to ```Misbehaving()``` might go unobserved, because the score may have already been evaluated before the other thread bumped it. This is not possible in with the current behavior because of cs_main: once we're in ```FinalizeNode```, we're guaranteed that no other thread is accessing it (unless they've cached the ```CNodeState*```, which would be a bug).

Obviously we don't have any threads doing that currently, so I'm wondering if we want to continue to enforce that invariant with a:
```c++
// This isn't a guarantee but it's close enough
assert(peer.use_count() <= 1);
```

I'm not too bothered if there's a rare misbehaving bump that gets missed. My concern is that Peer's destructor may eventually gain some functionality, and we'll want to know if we need to keep track of which thread deletes it.

Reply:
Thanks @theuni. This is a very good point. I'll add a comment about not relying on this being the final state of `Peer` in a future commit.